tests/test_settings.py

```diff
@@ -128,6 +130,60 @@ def test_base_path_overrides_config_file_code_and_deps(
     assert settings.deps == expected
 
 
+OPTION_VALUES = {
+    "code": {"a", "b", "c"},
+    "deps": {"d", "e"},
+    "ignore-undeclared": {"f", "g"},
+    "ignore-unused": {"h", "i", "j"},
+}
+
+
+def multivalued_optargs_grid() -> Iterable[List[str]]:
+    """Create command-line option/argument combinations which
+    mix order and number of multivalued parameters."""
+
+    T = TypeVar("T")
+
+    def subsequence_pairs(
+        xs: Tuple[T, ...]
+    ) -> Iterable[Tuple[Tuple[T, ...], Tuple[T, ...]]]:
+        assert len(xs) >= 2
+        for i in range(1, len(xs)):
+            yield xs[:i], xs[i:]
+
+    option_partitions = [
+        {pair for seq in permutations(items) for pair in subsequence_pairs(seq)}
+        for items in OPTION_VALUES.values()
+    ]
+
+    opt_by_arg = {arg: opt for opt, argvals in OPTION_VALUES.items() for arg in argvals}
+
+    for param_grid in set(product(*option_partitions)):
+        xss = list(chain(*param_grid))
+        random.shuffle(xss)
+        yield list(chain(*[[f"--{opt_by_arg[xs[0]]}"] + list(xs) for xs in xss]))
+
+
+@pytest.mark.parametrize("optargs", multivalued_optargs_grid())
+def test_multivalued_options_are_aggregated_correctly(optargs):
+    settings = run_build_settings(optargs)
+    assert settings.code == to_path_set(OPTION_VALUES["code"])
+    assert settings.deps == to_path_set(OPTION_VALUES["deps"])
+    assert settings.ignore_undeclared == set(OPTION_VALUES["ignore-undeclared"])
+    assert settings.ignore_unused == set(OPTION_VALUES["ignore-unused"])
+
+
+@pytest.mark.parametrize(
+    "optname",
+    set(act.dest for act in build_parser()._actions)  # pylint: disable=protected-access
+    & set(Settings.__fields__),
+)
+def test_settings_members_are_absent_from_namespace_if_not_provided_at_cli(optname):
+    parsed_cli = build_parser().parse_args([])
+    with pytest.raises(AttributeError):
+        getattr(parsed_cli, optname)
```

Comment:
Nit: I think this is more readable, if I have not overlooked any details:
```suggestion
    assert not hasattr(parsed_cli, optname)
```

Reply:
This one I'm going to leave, as I think the `raises` assertion is a bit stronger, like the difference between the membership check with `in` on a `Mapping` vs. the direct `[]` accessor. Most of the time the same, but with the `defaultdict`, only one would fail. I want to make the stronger type of assertion with this test.